resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -829,50 +968,176 @@ void add(DnsRecord r) {
+                    // the smallest for caching.
+                    serverName.update(parent.newRedirectServerAddress(resolved), r.timeToLive());
+                    return;
+                }
+                serverName = serverName.next;
+            }
+        }
+
+        // Now handle all AuthoritativeNameServer for which we had no ADDITIONAL record
+        void handleWithoutAdditionals(DnsNameResolver parent, DnsCache cache) {
+            AuthoritativeNameServer serverName = head;
+
+            while (serverName != null) {
+                if (serverName.address == null) {
+                    // Try to resolve via cache
+
+                    List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
+                    if (entries == null || entries.isEmpty()) {
+                        // These will be resolved on the fly if needed.
+                        serverName.address = InetSocketAddress.createUnresolved(
+                                serverName.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+                    } else {
+                        // We have some entries in the cache.
+                        for (int i = 0; i < entries.size(); i++) {
+                            InetAddress address = entries.get(i).address();
+                            if (address != null) {
+                                if (serverName.address != null) {
+                                    AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                                    server.next = serverName.next;
+                                    serverName.next = server;
+                                    serverName = server;
+
+                                    resolvedCount++;
+                                }
+                                // Use a TTL of Long.MIN_VALUE as we don't want to cache the actual address later on as
+                                // we have no idea about how long the TTL should be. When we encounter Long.MIN_VALUE
+                                // later on we will cache the unresolved address.
+                                serverName.update(parent.newRedirectServerAddress(address), Long.MIN_VALUE);
+                            }
+                        }
+                    }
                 }
                 serverName = serverName.next;
             }
-            return null;
         }
 
-        int size() {
-            return count;
+        /**
+         * Returns {@code true} if empty, {@code false} otherwise.
+         */
+        boolean isEmpty() {
+            return resolvedCount == 0;
+        }
+
+        /**
+         * Creates a new {@link List} which holds the {@link InetSocketAddress}es.
+         */
+        List<InetSocketAddress> cacheAndReturnAsList(AuthoritativeDnsServerCache cache, EventLoop loop) {
+            List<InetSocketAddress> addressList = new ArrayList<InetSocketAddress>(resolvedCount);
```

Comment:
@Scottmitch `resolvedCount` should be at least as large as we need or even larger as we update it also when we add entries as part of resolving these from the `DnsCache`. Can you tell me why you think this is not the case ?

Reply:
`handleWithoutAdditionals` will set `serverName.address` to an unresolved `InetSocketAddress` if no corresponding entries can be found in the cache without incrementing `resolvedCount`. IIUC the intention is to include elements which otherwise cannot be resolved (via additionals or cache) to avoid intermediate lookups in case of TTL differences. In this sense the `resolvedCount` seems to under count, and the "resolved" terminology seems to not account for everything we are interested in.

Reply:
ah sorry you are right... just renamed to servers and ensure we always update when we add a server address.